xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+
+Cross-Platform Support
+----------------------
+
+AppArmor only works in Linux, and cannot run on Linux guest Docker containers on a macOS host machine. This makes it difficult for developers running macOS to develop and test features related to sandboxed code execution. WebAssembly can run on any major operating system.
+
+Decisions
+*********
+
+#. We will create a new library in a new repo (``wasm-safe-exec``) for safely executing instructor-authored code using WebAssembly as the execution mechanism.
+
+   * The library will use ``wasmtime-py`` for Wasm code execution.
+   * The library must be runnable in edx-platform, meaning that it still needs to support Python 3.8 until edx-platform has been upgraded to 3.11.
+   * The library must be able to simultaneously support multiple code execution runtime environments.
+   * The library must not make ``edx-platform`` specific assumptions–e.g. ``UsageKeys``. It should be generally reusable for untrusted code execution by other projects.
+
+#. We will use the WebAssembly Component Model and create a Wasm Interface Type (WIT) file defining the interface between the host library and guest code execution runtime environments.
+
+   * It should be possible for third parties to easily create their own runtime environments.
+
+#. We will create two simple runtimes to start: Python 3.11 and JavaScript ES14.
+
+   * Runtime binaries will be created with `componentize-py <https://component-model.bytecodealliance.org/language-support/python.html>`_ for Python and and `jco <https://component-model.bytecodealliance.org/language-support/javascript.html>`_ for JavaScript.
+   * Python 3.11 is the first version of Python that officially supports compiling to ``wasm32-wasi``, which is why it's the one we'll start with.
+   * The Python runtime must be able to accept a course's ``python_lib.zip`` file.
+   * Parity with codejail is not feasible at this time because only a handful of our popular libraries are available. However, we should include numpy (`using Fermyon's instructions <https://www.fermyon.com/blog/introducing-componentize-py>`_) and whatever pure Python libraries are practical to do so.
+   * Preamble code (e.g. certain common library imports) will be the responsibility of the runtime environments.
+   * For this first iteration, the Wasm runtime binary build process can simply be documented, and does not need to be part of any package or CI. In particular, we do *not* want to add 100 MB+ ``.wasm`` files as a direct dependency of edx-platform.
```

Comment:
In the future plan, Do we have specific plans in mind about generateing the binaries easily? e.g. 

- Do we plan to enhance open edX's [cookie cutter](https://github.com/openedx/edx-cookiecutters) to support building these libraries?
- Or would it be part of `wasm-safe-exec`?

There could be any other options as well but I just wanted to point to the idea.